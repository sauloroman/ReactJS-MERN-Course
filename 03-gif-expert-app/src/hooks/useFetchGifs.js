import { useEffect, useState } from "react"
import { getGifs } from "../helpers/getGifs"

export const useFetchGifs = ( search ) => {

  const [result, setResult] = useState({
    data: null,
    hasError: null,
    isLoading: false,
  })

  useEffect(() => {
    getData( search )
  }, [])

  const getData = async () => {

    setResult({
      ...result,
      isLoading: true,
    })

    try {

      const data = await getGifs( search )

      setResult({
        data,
        hasError: null,
        isLoading: false,
      })
      
    } catch (error) {
      setResult({
        data: null,
        hasError: error,
        isLoading: false,
      })
    }

  }

  return {
    ...result,
  }

}