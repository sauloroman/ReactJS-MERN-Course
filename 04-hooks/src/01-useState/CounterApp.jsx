import { useState } from 'react';

export const CounterApp = () => {
  const [{ counter1, counter2, counter3 }, setCounter] = useState({
    counter1: 10,
    counter2: 20,
    counter3: 30,
  });

  return (
    <>
      <h1 className="heading">UseState &mdash; CounterApp</h1>

      <div className="counters">
        <div className="counter">
          <h2 className="counter__value" data-testid="counter-1">{counter1}</h2>
          <span className="counter__name">Contador 1</span>
          <button
            onClick={() =>
              setCounter((current) => ({
                ...current,
                counter1: counter1 + 1,
              }))
            }
            className="counter__btn"
          >
            Aumentar 1
          </button>
        </div>
        <div className="counter">
          <h2 className="counter__value" data-testid="counter-2">{counter2}</h2>
          <span className="counter__name">Contador 2</span>
          <button
            onClick={() =>
              setCounter((current) => ({
                ...current,
                counter2: counter2 + 1,
              }))
            }
            className="counter__btn"
          >
            Aumentar 2
          </button>
        </div>
        <div className="counter">
          <h2 className="counter__value" data-testid="counter-3">{counter3}</h2>
          <span className="counter__name">Contador 3</span>
          <button
            onClick={() =>
              setCounter((current) => ({
                ...current,
                counter3: counter3 + 1,
              }))
            }
            className="counter__btn"
          >
            Aumentar 3
          </button>
        </div>
      </div>
    </>
  );
};
