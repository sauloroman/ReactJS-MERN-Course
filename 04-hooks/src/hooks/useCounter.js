import { useState } from "react"

export const useCounter = ( initialValue = 1 ) => {

  const [counter, setCounter] = useState(initialValue)

  const onIncrementCounter = ( value = 1 ) => {
    setCounter( counter => counter + value )
  }

  const onDecrementCounter = ( value = 1 ) => {
    if ( counter === 0 ) return
    setCounter( counter => counter - value )
  }

  const onResetCounter = () => setCounter(initialValue)

  return {
    // State
    counter,

    // Methods
    onIncrementCounter,
    onDecrementCounter,
    onResetCounter
  }

}