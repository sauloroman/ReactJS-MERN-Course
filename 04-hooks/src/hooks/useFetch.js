import { useEffect } from "react"
import { useState } from "react"

export const useFetch = ( url = '' ) => {

  const [state, setState] = useState({
    data: null,
    hasError: null,
    isLoading: false,
  })

  useEffect(() => {
    getData()
  }, [ url ])

  const getData = async () => {

    setState({
      ...state,
      isLoading: true,
    })

    try {
      
      const res = await fetch( url );
      const data = await res.json()
    
      setState({
        ...state,
        data: data,
        isLoading: false,
      })

    } catch (error) {
      setState({
        ...state,
        hasError: error,
        isLoading: false,
      })
    }

  }

  return {
    data: state.data,
    hasError: state.hasError,
    isLoading: state.isLoading,
  }

}