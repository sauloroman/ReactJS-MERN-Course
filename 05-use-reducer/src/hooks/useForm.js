import { useState } from "react"

export const useForm = ( initialState = {} ) => {

  const [formState, setFormState] = useState(initialState)

  const onInputChange = ({ target: { name, value }}) => {
    setFormState({
      ...formState,
      [name]: value
    })
  }

  const onResetForm = () => setFormState( initialState )

  return {
    formState,
    ...formState,

    onInputChange,
    onResetForm,
  }

}