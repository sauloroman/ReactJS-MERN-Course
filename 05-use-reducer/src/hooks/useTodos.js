import { useReducer } from "react"
import { todoReducer } from "../helpers/todo.reducer"

const init = () => JSON.parse(localStorage.getItem("todos")) || []

export const useTodos = () => {
    const [todos, dispatch] = useReducer( todoReducer, [], init )
  
    const onAddTodo = function( newTodo = {} ) {
      const actionAdd = {
        type: '[TODO] - Add New todo',
        payload: newTodo,
      }
      dispatch( actionAdd )
    }
  
    const onToggleTodo = function( todoId = '' ) {
      const actionToggle = {
        type: '[TODO] - Toggle todo',
        payload: todoId
      }
      dispatch( actionToggle )
    }
  
    const onDeleteTodo = function( todoId = '' ) {
      const actionDelete = {
        type: '[TODO] - Delete todo',
        payload: todoId,
      }
      dispatch( actionDelete )
    }

    return {
      todos,

      onAddTodo,
      onToggleTodo,
      onDeleteTodo,
    }
}