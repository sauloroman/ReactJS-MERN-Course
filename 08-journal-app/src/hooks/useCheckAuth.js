import { useEffect } from "react"
import { useDispatch, useSelector } from "react-redux"
import { onAuthStateChanged } from "firebase/auth"
import { FirebaseAuth } from "../firebase/config,"
import { login, logout } from "../store/auth"
import { startLoadingNotes } from "../store/journal/journal.thunk"

export const useCheckAuth = () => {
  const dispatch = useDispatch()
  const { status } = useSelector(state => state.auth)

  useEffect(() => {

    onAuthStateChanged( FirebaseAuth, async( user ) => {
      if ( !user ) return dispatch( logout() )

      const { uid, email, photoURL, displayName} = user
      dispatch( login({ uid, email, photoURL, displayName}) )
      dispatch( startLoadingNotes() )
    })

  }, [])

  return status
  
}