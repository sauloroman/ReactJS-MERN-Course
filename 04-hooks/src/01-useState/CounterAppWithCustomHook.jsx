import { useCounter } from '../hooks';

export const CounterAppWithCustomHook = () => {
  const { counter, onDecrementCounter, onIncrementCounter, onResetCounter } =
    useCounter(1);

  return (
    <>
      <h1 className="heading">UseState &mdash; CounterAppCustomHook</h1>

      <div className="counter counter--unique">
        <span data-testid="counter" className="counter__value">{counter}</span>
        <span className="counter__name">Contador</span>

        <div className="counter__buttons">
          <button onClick={() => onIncrementCounter()} className="counter__btn">+1</button>
          <button onClick={onResetCounter} className="counter__btn">Reset</button>
          <button onClick={() => onDecrementCounter()} className="counter__btn">-1</button>
        </div>
      </div>
    </>
  );
};
