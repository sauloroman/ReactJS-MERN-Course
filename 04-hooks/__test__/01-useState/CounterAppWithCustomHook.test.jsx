import { render, fireEvent, screen } from "@testing-library/react"
import { CounterAppWithCustomHook } from "../../src/01-useState/CounterAppWithCustomHook"
import { useCounter } from '../../src/hooks/useCounter'

jest.mock('../../src/hooks/useCounter.js')

describe('Tests in CounterAppWithCustomHook', () => {

  const counterValue = 1
  const onDecrementCounterMock = jest.fn()
  const onIncrementCounterMock = jest.fn()
  const onResetCounterMock = jest.fn()
  
  useCounter.mockReturnValue({
    counter: counterValue,
    onDecrementCounter: onDecrementCounterMock,
    onIncrementCounter: onIncrementCounterMock,
    onResetCounter: onResetCounterMock,
  })

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('It must match with snapshot', () => {
    const { container } = render(<CounterAppWithCustomHook />)
    expect( container ).toMatchSnapshot();
  })

  test('it must show the counter', () => {
    render(<CounterAppWithCustomHook />)
    const counterElem = screen.getByTestId('counter')
    expect( counterElem.textContent ).toBe(String( counterValue ))
  })

  test('It must call onIncrementCounter function when the appropiate button is pushed', () => {
    render(<CounterAppWithCustomHook />)

    const buttonIncrement = screen.getByRole('button', { name: '+1' } )
    fireEvent.click( buttonIncrement )

    expect( onIncrementCounterMock ).toHaveBeenCalled()
  })

  test('It must call onDecrementCounter function when the appropiate button is pushed', () => {

    render( <CounterAppWithCustomHook /> )

    const buttonDecrement = screen.getByRole('button', { name: '-1' } )
    fireEvent.click( buttonDecrement )

    expect( onDecrementCounterMock ).toHaveBeenCalled()
  })

  test('It must call onResetCounter function when the appropiate button is pushed', () => {

    render( <CounterAppWithCustomHook /> )

    const buttonReset = screen.getByRole('button', { name: 'Reset' } )
    fireEvent.click( buttonReset )

    expect( onResetCounterMock ).toHaveBeenCalled()
  })

})