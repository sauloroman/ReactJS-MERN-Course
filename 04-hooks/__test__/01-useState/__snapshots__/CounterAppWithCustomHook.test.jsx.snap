// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in CounterAppWithCustomHook It must match with snapshot 1`] = `
<div>
  <h1
    class="heading"
  >
    UseState â€” CounterAppCustomHook
  </h1>
  <div
    class="counter counter--unique"
  >
    <span
      class="counter__value"
      data-testid="counter"
    >
      1
    </span>
    <span
      class="counter__name"
    >
      Contador
    </span>
    <div
      class="counter__buttons"
    >
      <button
        class="counter__btn"
      >
        +1
      </button>
      <button
        class="counter__btn"
      >
        Reset
      </button>
      <button
        class="counter__btn"
      >
        -1
      </button>
    </div>
  </div>
</div>
`;
